flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTask.java

```diff
@@ -114,4 +137,81 @@ protected void createInputProcessor(
 			operatorChain,
 			setupNumRecordsInCounter(mainOperator));
 	}
+
+	@Override
+	public Future<Boolean> triggerCheckpointAsync(
+			CheckpointMetaData metadata,
+			CheckpointOptions options,
+			boolean advanceToEndOfEventTime) {
+
+		CompletableFuture<Boolean> resultFuture = new CompletableFuture<>();
+		mainMailboxExecutor.execute(
+			() -> {
+				try {
+					/**
+					 * Contrary to {@link SourceStreamTask}, we are not using here
+					 * {@link StreamTask#latestAsyncCheckpointStartDelayNanos} to measure the start delay
+					 * metric, but we will be using {@link CheckpointBarrierHandler#getCheckpointStartDelayNanos()}
+					 * instead.
+					 */
+					pendingCheckpointCompletedFutures.put(metadata.getCheckpointId(), resultFuture);
+					checkPendingCheckpointCompletedFuturesSize();
+					triggerSourcesCheckpoint(new CheckpointBarrier(metadata.getCheckpointId(), metadata.getTimestamp(), options));
+				}
+				catch (Exception ex) {
+					// Report the failure both via the Future result but also to the mailbox
+					pendingCheckpointCompletedFutures.remove(metadata.getCheckpointId());
+					resultFuture.completeExceptionally(ex);
+					throw ex;
+				}
+			},
+			"checkpoint %s with %s",
+			metadata,
+			options);
+		return resultFuture;
+	}
+
+	private void checkPendingCheckpointCompletedFuturesSize() {
+		while (pendingCheckpointCompletedFutures.size() > MAX_TRACKED_CHECKPOINTS) {
+			Long minCheckpointID = Collections.min(pendingCheckpointCompletedFutures.keySet());
```

Comment:
I think we can end up with `O(MAX_TRACKED_CHECKPOINTS ^ 2)` here, right?
Even though it shouldn't happen normally (if map is cleared by triggerOnBarrier/abort), I think it's better to cap it by either using `LinkedHashMap`, `TreeMap`, or just reducing 
`MAX_TRACKED_CHECKPOINTS` to something very small.

Reply:
Ehhhh... ok, but that's overengineering a bit :)